File: 32.最长有效括号.js
/*
 * @lc app=leetcode.cn id=32 lang=javascript
 *
 * [32] 最长有效括号
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function (s) {
    let maxLength = 0;
    let i = 0;
    while (s[i]) {
        let length = getMaxLength(s.slice(i));
        if (length > maxLength) {
            maxLength = length;
        }
        i++;
    }
    return maxLength;
};

function getMaxLength(s) {
    let stack = 0;
    let i = 0;
    let maxLength = 0;
    while (s[i]) {
        if (s[i] === '(') {
            stack++;
        } else {
            stack--;
        }
        if (stack < 0) {
            return maxLength;
        } else if (stack === 0) {
            maxLength = i + 1;
        }
        i++;
    }
    return maxLength;
}

// let res = longestValidParentheses("");
// console.log(res);
// @lc code=end

File: 32.最长有效括号.py
#
# @lc app=leetcode.cn id=32 lang=python3
#
# [32] 最长有效括号
#

# @lc code=start
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        maxans = 0
        dp = [0] * len(s)
        for i in range(1, len(s)):
            if s[i] == ')':
                if s[i-1] == '(':
                    dp[i] = (dp[i-2] if i >= 2 else 0) + 2
                elif i - dp[i-1] > 0 and s[i-dp[i-1]-1] == '(':
                    dp[i] = dp[i - 1] + (dp[i - dp[i - 1] - 2]
                                         if (i - dp[i - 1]) >= 2 else 0) + 2
                maxans = max(maxans, dp[i])
        return maxans
# @lc code=end

solution = Solution()
res =solution.longestValidParentheses(')(')
print(res)